File: kotlin-electron/karakum/annotations/annotateJsPlainObject.js
import ts from "typescript"

export default (node, context) => {
    if (
        context.isAnonymousDeclaration

        && ts.isIntersectionTypeNode(node)

        && node.parent
        && ts.isParameter(node.parent)
        && ts.isIdentifier(node.parent.name)
        && node.parent.name.text === "init"

        && (
            (
                node.parent.parent
                && ts.isMethodSignature(node.parent.parent)
                && node.parent.parent.name.text === "fetch"

                && node.parent.parent.parent
                && ts.isInterfaceDeclaration(node.parent.parent.parent)
                && node.parent.parent.parent.name.text === "Net"
            )
            || (
                node.parent.parent
                && ts.isMethodDeclaration(node.parent.parent)
                && node.parent.parent.name.text === "fetch"

                && node.parent.parent.parent
                && ts.isClassDeclaration(node.parent.parent.parent)
                && node.parent.parent.parent.name
                && node.parent.parent.parent.name.text === "Session"
            )
        )
    ) {
        return `@kotlinx.js.JsPlainObject`
    }

    return null
}

File: kotlin-electron/karakum/src/jsMain/kotlin/electron/karakum/annotations/annotateJsPlainObject.kt
package electron.karakum.annotations

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.AnnotationContext
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.*

fun annotateJsPlainObject(node: Node, context: AnnotationContext) = nullable {
    ensure(context.isAnonymousDeclaration)

    ensure(isIntersectionTypeNode(node))

    val parameter = ensureNotNull(node.getParentOrNull())
    ensure(isParameter(parameter))

    val parameterName = ensureNotNull(parameter.name)
    ensure(isIdentifier(parameterName))
    ensure(parameterName.text == "init")

    nullable {
        val method = ensureNotNull(parameter.getParentOrNull())
        ensure(isMethodSignature(method))

        val methodName = method.name
        ensure(isIdentifier(methodName))
        ensure(methodName.text == "fetch")

        val interfaceNode = ensureNotNull(method.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))
        ensure(interfaceNode.name.text == "Net")

        "@kotlinx.js.JsPlainObject"
    } ?: nullable {
        val method = ensureNotNull(parameter.getParentOrNull())
        ensure(isMethodDeclaration(method))

        val methodName = method.name
        ensure(isIdentifier(methodName))
        ensure(methodName.text == "fetch")

        val classNode = ensureNotNull(method.getParentOrNull())
        ensure(isClassDeclaration(classNode))
        ensure(classNode.name?.text == "Session")

        "@kotlinx.js.JsPlainObject"
    }
}
